pkg/processor/procjs/jsprocessor.go

```diff
@@ -0,0 +1,184 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procjs
+
+import (
+	"context"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/dop251/goja"
+	"github.com/rs/zerolog"
+)
+
+const (
+	entrypoint = "process"
+)
+
+// JSProcessor is able to run processors defined in JavaScript.
+type JSProcessor struct {
+	runtime  *goja.Runtime
+	function goja.Callable
+}
+
+func New(src string, logger zerolog.Logger) (*JSProcessor, error) {
+	p := &JSProcessor{}
+	err := p.initJSRuntime(logger)
+	if err != nil {
+		return nil, cerrors.Errorf("failed initializing JS runtime: %w", err)
+	}
+
+	err = p.initFunction(src)
+	if err != nil {
+		return nil, cerrors.Errorf("failed initializing JS function: %w", err)
+	}
+
+	return p, nil
+}
+
+func (p *JSProcessor) initJSRuntime(logger zerolog.Logger) error {
+	rt := goja.New()
+	runtimeHelpers := map[string]interface{}{
+		"logger":  &logger,
+		"Record":  p.jsRecord,
+		"RawData": p.jsContentRaw,
+	}
+
+	for name, helper := range runtimeHelpers {
+		if err := rt.Set(name, helper); err != nil {
+			return cerrors.Errorf("failed to set helper %q: %w", name, err)
+		}
+	}
+
+	p.runtime = rt
+	return nil
+}
+
+func (p *JSProcessor) initFunction(src string) error {
+	prg, err := goja.Compile("", src, false)
+	if err != nil {
+		return cerrors.Errorf("failed to compile script: %w", err)
+	}
+
+	_, err = p.runtime.RunProgram(prg)
+	if err != nil {
+		return cerrors.Errorf("failed to run program: %w", err)
+	}
+
+	tmp := p.runtime.Get(entrypoint)
+	entrypointFunc, ok := goja.AssertFunction(tmp)
+	if !ok {
+		return cerrors.Errorf("failed to get entrypoint function %q", entrypoint)
+	}
+
+	p.function = entrypointFunc
+	return nil
+}
+
+func (p *JSProcessor) jsRecord(goja.ConstructorCall) *goja.Object {
+	// TODO accept arguments
+	// We return a record.Record struct, however because we are
+	// not changing call.This instanceof will not work as expected.
+
+	r := record.Record{
+		Metadata: make(map[string]string),
+	}
+	// We need to return a pointer to make the returned object mutable.
+	return p.runtime.ToValue(&r).ToObject(p.runtime)
+}
+
+func (p *JSProcessor) jsContentRaw(goja.ConstructorCall) *goja.Object {
+	// TODO accept arguments
+	// We return a record.RawData struct, however because we are
+	// not changing call.This instanceof will not work as expected.
+
+	r := record.RawData{}
+	// We need to return a pointer to make the returned object mutable.
+	return p.runtime.ToValue(&r).ToObject(p.runtime)
+}
```

Comment:
TBH I'd prefer if these functions were separate from the processor struct. Previously they were static, but maybe it would be best to gather them on a struct like `Helper` or something, to make it clear they are here to be called from the JS script.

Reply:
I actually had all of the JS stuff in a separate package, a "JS engine" which had nothing to do with processing itself. However, I realized that in that case the JS processor isn't doing much, so it felt like a premature refactoring.